app/src/main/java/com/jerboa/ui/components/common/ImageViewerDialog.kt

```diff
@@ -0,0 +1,153 @@
+import net.engawapg.lib.zoomable.zoomable
+
+val backColor = Color.Black
+val backColorTranslucent = Color.Black.copy(alpha = 0.4f)
+const val backFadeTime = 300
+
+@Composable
+fun ImageViewerDialog(url: String, onBackRequest: () -> Unit) {
+    @Composable
+    fun BarIcon(icon: ImageVector, name: String, modifier: Modifier = Modifier, onTap: () -> Unit) {
+        Box(
+            Modifier
+                .size(40.dp)
+                .clickable(onClick = onTap)
+                .then(modifier),
+        ) {
+            Icon(
+                modifier = Modifier.align(Alignment.Center),
+                imageVector = icon,
+                tint = Color.White,
+                contentDescription = name,
+            )
+        }
+    }
+
+    var showTopBar by remember { mutableStateOf(true) }
+
+    val topBarAlpha = animateFloatAsState(
+        targetValue = if (showTopBar) 1f else 0f,
+        animationSpec = tween(backFadeTime),
+    )
+    val backgroundColor = animateColorAsState(
+        targetValue = if (showTopBar) backColorTranslucent else backColor,
+        animationSpec = tween(backFadeTime),
+    )
+
+    val context = LocalContext.current
+    val coroutineScope = rememberCoroutineScope()
+
+    Dialog(
+        onDismissRequest = onBackRequest,
+        properties = DialogProperties(
+            dismissOnBackPress = true,
+            dismissOnClickOutside = false,
+            usePlatformDefaultWidth = false,
+        ),
+    ) {
+        Box(Modifier.background(backgroundColor.value)) {
+            Image(
+                painter = rememberAsyncImagePainter(url),
+                contentDescription = null,
+                modifier = Modifier
+                    .fillMaxSize()
+                    .zoomable(
+                        zoomState = rememberZoomState(),
+                        onTap = { showTopBar = !showTopBar },
+                    ),
+            )
+
+            Row(
+                modifier = Modifier
+                    .fillMaxWidth()
+                    .alpha(topBarAlpha.value)
+                    .padding(horizontal = 10.dp),
+                verticalAlignment = Alignment.CenterVertically,
+            ) {
+                BarIcon(icon = Icons.Filled.ArrowBack, name = "Back") {
+                    onBackRequest()
+                }
+
+                Spacer(Modifier.weight(1f))
+
+                BarIcon(icon = Icons.Outlined.Download, name = "Download") {
+                    coroutineScope.launch {
+                        SaveImage(url, context)
+                    }
+                }
+            }
+        }
+    }
+}
+
+suspend fun SaveImage(url: String, context: Context) {
+    val request = ImageRequest.Builder(context)
+        .data(url)
+        .crossfade(true)
+        .target(
+            onSuccess = {
+                val fileName = Uri.parse(url).pathSegments.last()
+                saveBitmap(context, it.toBitmap(), Bitmap.CompressFormat.WEBP, "image/webp", fileName)
```

Comment:
I've gone back and forth on webp... it is smaller and that's nice, but also according to the dev of a picture library I really like (photoprism), google's support is terrible and nearly all open-source devs that work with picture software hate it. Plus people will probably want to download the high-res originals, in whatever format the pictrs server provides.

I'm fairly sure the imageUrl I'm providing can be jpg, jpeg, png, or webp for the thumbnails only, so I spose just match whatever's there.

Reply:
Yeah it's not that great in terms of support, even in Go which is made by Google the support is a bit iffy. What I wanted to do initially was to just download the image straight to a file, but it seems like there's no way to get the bitmap's raw data. I'll see what I can do.